File: EmarsysPredictShop/EmarsysPredictShop/ItemDetailViewController.m
//
//  ItemDetailViewController.m
//  EmarsysPredictShop
//

@import Foundation;
#import <EmarsysPredictSDK/EmarsysPredictSDK.h>

#import "ItemDetailViewController.h"
#import "Cart+EmarsysPredictSDKExtensions.h"
#import "RecommendedItem.h"

@interface ItemDetailViewController () <RecommendationManagerDelegate>

@end

@implementation ItemDetailViewController

static NSString *const reuseIdentifier = @"ItemCell";

- (void)viewDidLoad {
    [super viewDidLoad];

    self.recommendationManager = [[RecommendationManager alloc] init];
    self.recommendationManager.recommendResults = [[NSMutableArray alloc] init];
    self.recommendationManager.delegate = self;
    self.recommendedCollectionView.dataSource = self.recommendationManager;
    self.recommendedCollectionView.delegate = self.recommendationManager;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self updateItem];
}

- (void)updateWithItem:(Item *)newItem {
    self.item = newItem;
    [self updateItem];
}

- (void)updateItem {
    self.name.text = self.item.title;
    self.availability.text = [NSString
        stringWithFormat:@"Availability: %@",
                         self.item.available ? @"IN STOCK" : @"OUT OF STOCK"];
    self.availability.textColor =
        self.item.available
            ? [UIColor colorWithRed:0.0 green:0.8 blue:0.0 alpha:1.0]
            : [UIColor colorWithRed:0.8 green:0.0 blue:0.0 alpha:1.0];
    self.price.text = [NSString stringWithFormat:@"$%.02f", self.item.price];

    NSURL *url = [NSURL URLWithString:_item.image];

    NSURLSessionTask *task = [[NSURLSession sharedSession]
          dataTaskWithURL:url
        completionHandler:^(NSData *_Nullable data,
                            NSURLResponse *_Nullable response,
                            NSError *_Nullable error) {
          if (!data) {
              return;
          }
          UIImage *image = [UIImage imageWithData:data];
          if (!image) {
              return;
          }
          dispatch_async(dispatch_get_main_queue(), ^{
            [self.image setImage:image];
          });
        }];

    [task resume];

    [self sendRecommend];
}

- (IBAction)addToCart:(id)sender {
    [[Cart sharedCart] addItem:_item];
    [[NSNotificationCenter defaultCenter] postNotificationName:@"CartChanged"
                                                        object:nil];
    UIAlertController *alert = [UIAlertController
        alertControllerWithTitle:@"Item added to cart"
                         message:nil
                  preferredStyle:UIAlertControllerStyleAlert];

    UIAlertAction *defaultAction =
        [UIAlertAction actionWithTitle:@"OK"
                                 style:UIAlertActionStyleDefault
                               handler:^(UIAlertAction *action){
                               }];
    [alert addAction:defaultAction];
    [self presentViewController:alert animated:YES completion:nil];

    [self sendRecommend];
}

- (void)sendRecommend {
    [self.recommendationManager.recommendResults removeAllObjects];
    [self.recommendedCollectionView reloadData];

    EMSession *emarsysSession = [EMSession sharedSession];
    EMTransaction *transaction =
        [[EMTransaction alloc] initWithSelectedItemView:_item.srcItem];

    NSArray<EMCartItem *> *cartItems = [[Cart sharedCart] convertItems];
    [transaction setCart:cartItems];

    // Passing on item ID to report product view. Item ID should match the
    // value listed in the Product Catalog
    [transaction setView:_item.itemID];

    NSString *logic = @"RELATED";

    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];

    NSString *userLogic = [defaults objectForKey:ItemDetailLogic];
    if (userLogic) {
        logic = userLogic;
    }

    EMRecommendationRequest *recommend =
        [[EMRecommendationRequest alloc] initWithLogic:logic];

    recommend.limit = 10;

    recommend.completionHandler = ^(EMRecommendationResult *_Nonnull result) {

      // Process result
      NSLog(@"%@", result.featureID);
      [result.products
          enumerateObjectsUsingBlock:^(EMRecommendationItem *_Nonnull obj,
                                       NSUInteger idx, BOOL *_Nonnull stop) {

            Item *item = [[Item alloc] initWithItem:obj];
            [self.recommendationManager.recommendResults addObject:item];
          }];

      dispatch_async(dispatch_get_main_queue(), ^{
        [self.recommendedCollectionView reloadData];
      });
    };
    [transaction recommend:recommend];

    // Firing the transaction. Should be the last call on the page, called only
    // once.
    [emarsysSession sendTransaction:transaction
                       errorHandler:^(NSError *_Nonnull error) {
                         // Some error happened
                         NSLog(@"value: %@", error);
                         return;
                       }];
}

@end

File: EmarsysPredictShopSwift/EmarsysPredictShopSwift/ItemDetailViewController.swift
//
//  ItemDetailViewController.swift
//  EmarsysPredictShopSwift
//
//

import UIKit
import EmarsysPredictSDK

class ItemDetailViewController: UIViewController, RecommendationManagerDelegate {
    
    @IBOutlet weak var image: UIImageView!
    @IBOutlet weak var name: UILabel!
    @IBOutlet weak var availability: UILabel!
    @IBOutlet weak var price: UILabel!
    @IBOutlet weak var recommendedCollectionView: UICollectionView!
    
    var item: Item!
    var recommendationManager: RecommendationManager!
    
    @IBAction func addToCart(sender: UIButton) {
        Cart.sharedCart.addItem(item)
        NSNotificationCenter.defaultCenter().postNotificationName("CartChanged", object: nil)
        
        let alert = UIAlertController(title: "Item added to cart", message: nil, preferredStyle: .Alert)
        let defaultAction = UIAlertAction(title: "OK", style: UIAlertActionStyle.Default, handler: {(action) in })
        alert.addAction(defaultAction)
        presentViewController(alert, animated: true, completion: nil)
        
        sendRecommend()
    }
    
    func updateWithItem(item: Item) {
        self.item = item
        updateItem()
    }
    
    func updateItem() {
        name.text = item.title
        
        let available: String!
        let color: UIColor!
        
        if item.available ?? false {
            available = "IN STOCK"
            color = UIColor(red: 0.0, green: 0.8, blue: 0.0, alpha: 1.0)
        } else {
            available = "OUT OF STOCK"
            color = UIColor(red: 0.8, green: 0.0, blue: 0.0, alpha: 1.0)
        }
        
        availability.text = String(format: "Availability: %@", available)
        availability.textColor = color
        price.text = String(format: "$%.02f", item.price!)
        
        let url = NSURL(string: item.image!)
        let session = NSURLSession.sharedSession()
        session.dataTaskWithURL(url!, completionHandler: {(data, response, error) in
            guard let data = data else { return }
            guard let image = UIImage(data: data) else { return }
            dispatch_async(dispatch_get_main_queue()) {
                self.image.image = image
            }
        }).resume()
        
        sendRecommend()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        recommendationManager = RecommendationManager()
        recommendationManager.delegate = self
        recommendedCollectionView.dataSource = recommendationManager
        recommendedCollectionView.delegate = recommendationManager
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        updateItem()
    }
    
    func sendRecommend() {
        recommendationManager.recommendResults.removeAll()
        self.recommendedCollectionView.reloadData()
        
        let emarsysSession = EMSession.sharedSession()
        let transaction = EMTransaction.init(item: item.srcItem)
        let cartItems = Cart.sharedCart.convertItems()
        transaction.setCart(cartItems)
        transaction.setView(item.itemID)
        
        var logic = "RELATED"
        
        let defaults = NSUserDefaults.standardUserDefaults()
        if let userLogic = defaults.objectForKey(UserDefaultsKeys.DetailLogic.rawValue) {
            logic = userLogic as! String
        }
        
        let recommend = EMRecommendationRequest(logic: logic)
        
        recommend.limit = 10
        recommend.completionHandler = {result in
            // Process result
            for obj in result.products {
                let item = Item(withItem: obj)
                self.recommendationManager.recommendResults.append(item)
            }
            
            dispatch_async(dispatch_get_main_queue()) {
                self.recommendedCollectionView.reloadData()
            }
        }
        transaction.recommend(recommend)
        
        // Firing the transaction. Should be the last call on the page, called only once.
        emarsysSession.sendTransaction(transaction, errorHandler: {(error) in
            // Some error happened
            NSLog("value: %@", error)
            return
        })
    }
    
}
